crates/astria-grpc-mock/src/mock_server.rs

```diff
@@ -0,0 +1,255 @@
+            mock_id,
+            server_state: self.state.clone(),
+        }
+    }
+
+    pub async fn verify(&self) {
+        debug!("verifying mock expectations");
+        if let VerificationOutcome::Failure(failed_verifications) = self.state.read().await.verify()
+        {
+            let received_requests_message =
+                if let Some(received_requests) = &self.state.read().await.received_requests {
+                    if received_requests.is_empty() {
+                        "The server did not receive any request.".into()
+                    } else {
+                        received_requests.iter().enumerate().fold(
+                            "Received requests:\n".to_string(),
+                            |mut message, (index, request)| {
+                                _ = writeln!(message, "- Request #{}", index + 1,);
+                                _ = request.print(&mut message);
+                                message
+                            },
+                        )
+                    }
+                } else {
+                    "Enable request recording on the mock server to get the list of incoming \
+                     requests as part of the panic message."
+                        .into()
+                };
+            let verifications_errors: String =
+                failed_verifications.iter().fold(String::new(), |mut s, m| {
+                    _ = writeln!(s, "- {}", m.error_message());
+                    s
+                });
+            let error_message = format!(
+                "Verifications failed:\n{verifications_errors}\n{received_requests_message}",
+            );
+            if std::thread::panicking() {
+                debug!("{}", &error_message);
+            } else {
+                panic!("{}", &error_message);
+            }
+        }
+    }
+}
+
+impl Drop for MockServer {
+    // Clean up when the `MockServer` instance goes out of scope.
+    fn drop(&mut self) {
+        futures::executor::block_on(self.verify());
+        // The sender half of the channel, `shutdown_trigger`, gets dropped here
+        // Triggering the graceful shutdown of the server itself.
+    }
+}
+
+pub struct MockGuard {
+    notify: Arc<(Notify, AtomicBool)>,
+    mock_id: MockId,
+    server_state: Arc<RwLock<MockServerState>>,
+}
+
+impl MockGuard {
+    pub async fn wait_until_satisfied(self) {
+        let (notify, flag) = &*self.notify;
+        let mut notification = pin!(notify.notified());
+
+        // listen for events of satisfaction.
+        notification.as_mut().enable();
+
+        // check if satisfaction has previously been recorded
+        if flag.load(std::sync::atomic::Ordering::Acquire) {
+            return;
+        }
+
+        // await event
+        notification.await;
+    }
+}
+
+struct MockRequest {
+    inner: AnyRequest,
+}
+
+impl MockRequest {
+    fn print(&self, mut buffer: impl std::fmt::Write) -> std::fmt::Result {
```

Comment:
wiremock has a limiter for the number of chars each report should print. I have removed that config for this PR but I am inclined to bring it back later (because they have probably arrived there for a good reason). It doesn't match very well with `Display`.